[종근] QnA 2단계(로그인 상태) 구현완료

Bellroute opened the pull request:
- QuestionAcceptanceTest 구현
- QuestionController 구현
- QnaServiceTest 구현
- QnaService의 질문 CRUD기능 구현

기존에 구현된 코드를 뒤늦게 이해하는 바람에 미션 처음에 잘못 구현했던 Atdd를 기능 구현 후에 다시 수정하게 되었습니다.

Bellroute committed:
feat: QuestionAcceptanceTest

Bellroute committed:
feat: QuestionController

Bellroute committed:
feat: UserService 단위테스트

Bellroute committed:
feat: QnaService 질문기능 구현

Bellroute committed:
feat: QuestionDTO

Bellroute committed:
fix: 인수테스트, 서비스단위테스트 수정

SimKiSeong commented on src/test/java/support/test/AcceptanceTest.java:
```diff
@@ -36,6 +42,10 @@ protected User defaultUser() {
         return findByUserId(DEFAULT_LOGIN_USER);
     }
 
+    protected Question defaultQuestion() {
```
acceptance Test는 인수테스트로써 저희가 원하는 상황을 만들고 
그 상황에 맞는 테스트를 하기 위해 사용하는 것으로 알고 있습니다.
이번 QnA 게시판에서는 로그인이 되어 있는 상황을 만들기 위해 acceptance Test를 사용하고 있습니다.
즉 acceptance Test안에 있는 모든 변수나 함수는 로그인 상황을 만들기 위해 구현된 코드라고 생각합니다.
그렇기 때문에 지금 구현되어 있는 Question관련 코드들은 이곳에 적절하다고 생각하지 않습니다.
Question컨트롤러 같은곳에서 불러오는게 어떨까요?

wijehyeon replied:
흠 저도 여기에 제 임의로 코드를 넣었는데 생각해봐야겠군요

Bellroute replied:
좋은 지적 감사합니다

dadadamarine replied:
AcceptanceTest는 말씀해주신대로 로그인 인증을 테스트에서 건너뛰기위해 만들어진 테스트입니다.
다른 Question에 대한 default값을 설정해주고싶다면 관련 Test클래스를 만들어서 지정해주는게 좋을것 같습니다.

hyukjin-lee replied:
AcceptanceTest 클래스는 인수테스트 구현시, 공통으로 필요한 부분에 대한걸 추상화 시켜서 만든 클래스입니다. 인수테스트는 실서비스의 특정 기능이 잘 도는지 테스트하는 것인데, 대부분의 기능들이 로그인을 요구하고 있으므로 이 부분을 공통 부분으로 추출해서 구현해놓은 추상 클래스가 AcceptanceTest입니다.

hyukjin-lee replied:
저는 이 부분이 Question 인수테스트 하는 곳에 필드로 있어도 괜찮을 것 같다는 생각을 해보네요

SimKiSeong reviewed (commented):
제 생각이 맞을 지는 모르겠지만 혹시 이와 관련된 생각이 있으신 분들은 꼭 댓글 달아주세요

wijehyeon commented on src/main/java/codesquad/dto/QuestionDTO.java:
```diff
@@ -0,0 +1,16 @@
+    private String title;
+    private String contents;
+
+    public QuestionDTO(String title, String contents) {
```
롬복을 사용하셨다면 annotation추가로 해당 생성자도 생성가능해보이네요

wijehyeon commented on src/main/java/codesquad/service/QnaService.java:
```diff
@@ -25,7 +27,8 @@
 
-    public Question create(User loginUser, Question question) {
+    public Question create(User loginUser, QuestionDTO questionDTO) {
+        Question question = new Question(questionDTO.getTitle(), questionDTO.getContents());
```
questionDTO만 넘겨주고 Question이 일을하게하면 좋을 것 같아요

wijehyeon replied:
근데 그렇게 하면 코드가 늘어나서 오히려 안좋나..?

Bellroute replied:
좋은 생각인 것 같습니다!

wijehyeon commented on src/main/java/codesquad/service/QnaService.java:
```diff
@@ -36,14 +39,19 @@ public Question create(User loginUser, Question question) {
-        return null;
+    public Question update(User loginUser, long id, QuestionDTO updatedQuestionDTO) {
+        Question question = questionRepository
+                .findById(id).orElseThrow(UnAuthorizedException::new);
```
해당하는 id의 Question을 못찾았을때는 NoSuchElementException아닐까요

wijehyeon commented on src/main/java/codesquad/service/QnaService.java:
```diff
@@ -36,14 +39,19 @@ public Question create(User loginUser, Question question) {
-        // TODO 삭제 기능 구현
+    public void deleteQuestion(User loginUser, long questionId) throws Exception {
+        Question question = questionRepository
+                .findById(questionId).orElseThrow(UnAuthorizedException::new);
```
흠 여기도??

wijehyeon commented on src/main/java/codesquad/service/QnaService.java:
```diff
@@ -63,4 +71,8 @@ public Answer deleteAnswer(User loginUser, long id) {
     }
+
+    public Question findById(User loginUser, long id) {
+        return null;
```
null?

Bellroute replied:
answer관련된 코드 밑에 있어서 확인을 못했네요

wijehyeon reviewed (commented):
👍

Bellroute committed:
fix: 리뷰반영
1. QuestionDTO에 lombok 어노테이션으로 생성자 구현
2. Question Optional의 예외를 NoSuchElementException으로 변경
3. defaultQuestion 구현 위치 변경
4. Question객체에 일 부여

esp2ar0 commented on src/main/java/codesquad/service/QnaService.java:
```diff
@@ -36,14 +37,23 @@ public Question create(User loginUser, Question question) {
+    public Question update(User loginUser, long id, QuestionDTO updatedQuestionDTO) {
+        Question question = questionRepository
+                .findById(id).orElseThrow(NoSuchElementException::new);
+        question.update(loginUser, updatedQuestionDTO);
```
도메인에 너무 많은 일을 떠넘기는게 아닌가 싶어요.
객체 자신에게 일을 시키는 것은 좋지만, 도메인은 가장 코어한 객체이므로 조심해서 사용해야 합니다.
모든 상황에 대처 가능하게. 즉 유연하게 코딩해보세요

dadadamarine commented on src/main/java/codesquad/web/QuestionController.java:
```diff
@@ -0,0 +1,59 @@
+    public String update(@LoginUser User loginUser, @PathVariable long id, QuestionDTO questionDTO) {
+        qnaService.update(loginUser, id, questionDTO);
+
+        return String.format("redirect:/questions/%d", id);
```
이렇게 사용하는것과 String의 + 연산을 사용하는것과 성능의 차이가 어떻게 나는지 궁금합니다.
+연산은 새 String클래스를 만드는것이라고 알고있는데
String.format은 다르게 동작하나요?

esp2ar0 replied:
성능보다는 그냥.. 가독성이랄까 별 상관 없다고 생각해요

dadadamarine commented on src/test/java/codesquad/service/QnaServiceTest.java:
```diff
@@ -0,0 +1,70 @@
+        user = new User("sanjigi", "password", "name", "javajigi@slipp.net");
+        question = new Question("title", "this is test");
+        question.writeBy(user);
+        System.out.println("유저" + user.getId());
```
출력 코드는 삭제해야합니다. + 저도 아직 안쓰긴 하는데 Logger 사용을 습관하하면 좋을것 같습니다.

dadadamarine commented on src/test/java/codesquad/service/QnaServiceTest.java:
```diff
@@ -0,0 +1,70 @@
+
+    @Test(expected = CannotDeleteException.class)
+    public void delete_failed_when_wrong_user() throws Exception {
+        qnaService.deleteQuestion(null, question.getId());
```
LoginUser 부분에 null을 넣어 주셨는데, 로그인 안된 유저가 오면 null값을 가지나요? 
실제 동작할 시에 null이 입력되어, 그 상황을 테스트한것인지 궁금합니다.

dadadamarine commented on src/test/java/codesquad/service/QnaServiceTest.java:
```diff
@@ -0,0 +1,70 @@
+    @Test
+    public void update_question() {
+        qnaService.update(user, question.getId(), new QuestionDTO("updateTitle", "this is update"));
+        assertThat(qnaService.findById(question.getId()).get().getContents(), is("this is update"));
```
테스트 정상동작하나요? 
현재 이 소스에서 사용된 user는 Repository를 통해 찾은 유저가 아니라
기존에 있는 sanjigi 유저와는 다른 객체로 보입니다 (현재 jpa가 쿼리를 보내기 전이므로 Id가 null상태로보임)
question.isOnwer메서드가 equals로 동작하던데 정상작동을 안하는게 맞아보입니다.
